### PR TITLE
add_hook

### PR DESCRIPTION
**Pull request by PubChimps**
Pushed most functionality from Operator to Hook, in test env getting No module named 'airflow.hooks.base'?

### CODE REVIEW

File: airflow-provider-fivetran/operators/fivetran.py [hunk 1]
```diff
@@ -3,11 +3,12 @@
 import time
 
 import requests
-import pendulum
 
 from airflow.models import BaseOperator
 from airflow.utils.decorators import apply_defaults
 
+from airflow-provider-fivetran.airflow-provider-fivetran.hooks.fivetran import FivetranHook
```
Comment (crw) [inline airflow-provider-fivetran/operators/fivetran.py:10]: I believe `.hooks.fivetran` should work. `airflow-provider-fivetran.airflow-provider-fivetran` is not going to work with the installed package.

File: airflow-provider-fivetran/operators/fivetran.py [hunk 2]
```diff
@@ -29,8 +30,11 @@ class FivetranOperator(BaseOperator):
...
         self,
-        api_key=None,
-        api_secret=None,
+        run_name: Optional[str] = None,
+        timeout_seconds: Optional[int] = None,
+        fivetran_conn_id: str = 'fivetran',
+        fivetran_retry_limit: int = 3,
+        fivetran_retry_delay: int = 1,
         connector_id=None,
         poll_status_every_n_seconds: int = 15,
```
Comment (crw) [inline airflow-provider-fivetran/operators/fivetran.py:39]: There is probably a better name for this, like `poll_interval` or `poll_interval_seconds`

File: airflow-provider-fivetran/operators/fivetran.py [hunk 3]
```diff
@@ -48,121 +52,24 @@ def __init__(
...
-
-        return {
-            "succeeded_at": succeeded_at.to_iso8601_string(),
-            "connector_id": self.connector_id,
-        }
+        hook = self._get_hook()
+        resp = hook.check_connector(self.connector_id)
+        resp = hook.set_manual_schedule(self.connector_id)
+        resp = hook.start_fivetran_sync(self.connector_id)
+        resp = hook.start_fivetran_sync(self.connector_id, self.poll_status_every_n_seconds)
```
Comment (crw) [inline airflow-provider-fivetran/operators/fivetran.py:75]: We are going to want to break this out into different methods.

**Review by crw — CHANGES_REQUESTED** [review]

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+            attempt_num += 1
+            sleep(self.retry_delay)
+
+    def _log_request_error(self, attempt_num: int, error: str) -> None:
+        self.log.error('Attempt %s API Request to Fivetran failed with reason: %s', attempt_num, error)
+
+    def check_connector(self, connector_id):
+        # Make sure connector configuration has been completed successfully and is not broken.
+
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:124]: method docs should be docstring e.g. 
"""Docs go here"""
Also, we are using a specific documentation methodology that includes documenting the params in an RST-friendly format (I am less familiar with this doc pattern, we need to do some research :)

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        log.info(
+            "Connector type: {}, connector schema: {}".format(
+                connector_details["service"], connector_details["schema"]
+            )
+        )
+        log.info("Connectors logs at {}".format(URL_LOGS))
+        return resp
+
+    def set_manual_schedule(self, connector_id):
+        # Set connector to manual sync mode, required to force sync through the API
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:156]: Docstring please!

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+import time
+
+
+from airflow import __version__
+from airflow.exceptions import AirflowException
+from airflow.hooks.base_hook import BaseHook
+
+log = logging.getLogger(__name__)
+
+class FivetranHook(BaseHook):  # noqa
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:19]: I have no idea what `# noqa` does, but we likely don't need it here. Probably a Databricks code coverage thing.

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        Utility function to perform an API call with retries
+        :param endpoint_info: Tuple of method and endpoint
+        :type endpoint_info: tuple[string, string]
+        :param json: Parameters for this API call.
+        :type json: dict
+        :return: If the api call returns a OK status code,
+            this function returns the response in JSON. Otherwise,
+            we throw an AirflowException.
+        :rtype: dict
+        """
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:63]: ^^^ This is an example of a well-formatted method docstring.

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        # Make sure connector configuration has been completed successfully and is not broken.
+        
+        resp = self._do_api_call(('GET',connector_id))
+        connector_details = resp["data"]
+        URL_LOGS = "https://fivetran.com/dashboard/connectors/{}/{}/logs".format(
+            connector_details["service"], connector_details["schema"]
+        )
+        URL_SETUP = "https://fivetran.com/dashboard/connectors/{}/{}/setup".format(
+            connector_details["service"], connector_details["schema"]
+        )
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:132]: Not critical, but we should switch to using f-strings. I am less familiar with this feature but they seem to be being used as the current standard of practice.

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        setup_state = connector_details["status"]["setup_state"]
+        if setup_state != "connected":
+            EXC_SETUP: str = (
+                'Fivetran connector "{}" not correctly configured, status: {}; '
+                + "please complete setup at {}"
+            )
+            raise ValueError(
+                EXC_SETUP.format(self.connector_id, setup_state, URL_SETUP)
+            )
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:141]: f-strings needed. :)

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        succeeded_at = self.parse_timestamp(connector_details["succeeded_at"])
+        failed_at = self.parse_timestamp(connector_details["failed_at"])
+        previous_completed_at = succeeded_at if succeeded_at > failed_at else failed_at
+        # URL for connector logs within the UI
+        log.info(
+            "Connector type: {}, connector schema: {}".format(
+                connector_details["service"], connector_details["schema"]
+            )
+        )
+        log.info("Connectors logs at {}".format(URL_LOGS))
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:152]: f-string?

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        return resp
+
+    def set_manual_schedule(self, connector_id):
+        # Set connector to manual sync mode, required to force sync through the API
+        return self._do_api_call(('PATCH',connector_id),json.dumps({"schedule_type": "manual"}))
+
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:158]: two line breaks between methods

Comment (crw) [reply airflow_provider_fivetran/hooks/fivetran.py:158]: We should filter all files through black -- I'll show you how to set this up when we talk. We can also make this a part of the commit CI process.

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        URL_SETUP = "https://fivetran.com/dashboard/connectors/{}/{}/setup".format(
+            connector_details["service"], connector_details["schema"]
+        )
+        setup_state = connector_details["status"]["setup_state"]
+        if setup_state != "connected":
+            EXC_SETUP: str = (
+                'Fivetran connector "{}" not correctly configured, status: {}; '
+                + "please complete setup at {}"
+            )
+            raise ValueError(
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:139]: For Astronomer: does it make sense to raise exceptions like this to kill an Operator and fail? If so, should we be raising specific Airflow exceptions? (I am guessing the answer is Yes :)

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+                EXC_SETUP.format(self.connector_id, setup_state, URL_SETUP)
+            )
+        # We need to know the previous job's completion time to know if the job succeeded or failed
+        succeeded_at = self.parse_timestamp(connector_details["succeeded_at"])
+        failed_at = self.parse_timestamp(connector_details["failed_at"])
+        previous_completed_at = succeeded_at if succeeded_at > failed_at else failed_at
+        # URL for connector logs within the UI
+        log.info(
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:147]: I *think* but am not sure, but I think we use `self.log` for all instances where we are currently calling `log`. I am guessing that attaches more log context.

File: airflow_provider_fivetran/hooks/fivetran.py (hunk 10 above)
Comment (crw) [reply airflow_provider_fivetran/hooks/fivetran.py:152]: self.log?

File: airflow_provider_fivetran/hooks/fivetran.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+            succeeded_at = self.parse_timestamp(current_details["succeeded_at"])
+            failed_at = self.parse_timestamp(current_details["failed_at"])
+            current_completed_at = (
+                succeeded_at if succeeded_at > failed_at else failed_at
+            )
+            # The only way to tell if a sync failed is to check if its latest failed_at value
+            # is greater than then last known "sync completed at" value.
+            if failed_at > previous_completed_at:
+                raise ValueError(
+                    'Fivetran sync for connector "{}" failed; please see logs at {}'.format(
```
Comment (crw) [inline airflow_provider_fivetran/hooks/fivetran.py:187]: f-string?

**Review by crw — CHANGES_REQUESTED** [review]

File: airflow-provider-fivetran/operators/fivetran.py (hunk 3 above)
Comment (pgzmnk) [reply airflow-provider-fivetran/operators/fivetran.py]: +1 on breaking this out into different methods

As you noted in the `poll_fivetran_sync` function's to-do, this function is a good candidate for a sensor. The [HTTP sensor](https://github.com/apache/airflow/blob/master/airflow/providers/http/sensors/http.py#L99) provides a clear, related example.

``` python
def poke(self, context: Dict[Any, Any]) -> bool:
        from airflow.utils.operator_helpers import make_kwargs_callable

        self.log.info('Poking: %s', self.endpoint)
        try:
            response = self.hook.run(
                self.endpoint,
                data=self.request_params,
                headers=self.headers,
                extra_options=self.extra_options,
            )
            if self.response_check:
                kwargs_callable = make_kwargs_callable(self.response_check)
                return kwargs_callable(response, **context)

        except AirflowException as exc:
            if str(exc).startswith("404"):
                return False

            raise exc

        return True
```

**Review by crw — APPROVED** [review]
Approving, I will make changes and submit a PR based on the feedback of this PR.

**Pull request merged**